### PR TITLE
Add a `/feed` endpoint that returns an ATOM feed. Fixes #99.

### LINKED ISSUE #99: Add an RSS/Atom feed
Add an RSS/Atom feed so that people can subscribe to the series as a podcast.

### PR DESCRIPTION
**Pull request by hashhar**

### CODE REVIEW

**Comment by hishamco** [comment]
@hashhar I think you missed the calling part to access this from UI

**Comment by hashhar** [comment]
@hishamco Oops. I'll add a link. Should I add it just to the Index view or the shared razor files (the header or footer maybe?)

File: src/live.asp.net/Controllers/HomeController.cs
Comment (hishamco) [inline src/live.asp.net/Controllers/HomeController.cs]: Why you set the `disableCache` false all the time?!! if that's your intend I don't think its needed in the constructor, you can passing it directly in `GetRecordedShowsAsync`

File: src/live.asp.net/Services/FeedService.cs
Comment (hishamco) [inline src/live.asp.net/Services/FeedService.cs]: Please move the return inside the curly braces

File: src/live.asp.net/Services/FeedService.cs
Comment (hishamco) [inline src/live.asp.net/Services/FeedService.cs]: Can we simplify producing the feed using POCO and `XmlSerializer`. This is just an idea to clean the code and make it more readable

**Review by hishamco — CHANGES_REQUESTED** [review]

File: src/live.asp.net/Controllers/HomeController.cs
Comment (hashhar) [reply src/live.asp.net/Controllers/HomeController.cs]: I was trying to use the same thing as in the `Index` action. I couldn't understand where `disableCache` was coming from.

File: src/live.asp.net/Services/FeedService.cs
Comment (hashhar) [reply src/live.asp.net/Services/FeedService.cs]: How? Can we pass our own schema to XMLSerializer? Sorry I haven't used XmlSerializer before.

**Comment by hishamco** [comment]
IMHO adding the feed links in the footer is the best option 😄

**Comment by hashhar** [comment]
I did just that. What about using POCO? I'll try and use it and push a commit when it works.

File: src/live.asp.net/Services/FeedService.cs
Comment (hishamco) [reply src/live.asp.net/Services/FeedService.cs]: My idea is to use POCO, `Rss`, `Channel` and `Item` classes after that we can use the `XmlSerializer` to serialize the `Rss` instance

File: src/live.asp.net/Views/Shared/_Layout.cshtml
Comment (hishamco) [inline src/live.asp.net/Views/Shared/_Layout.cshtml:33]: Nice catch 😄

File: src/live.asp.net/Controllers/HomeController.cs
Comment (hishamco) [reply src/live.asp.net/Controllers/HomeController.cs]: https://github.com/aspnet/live.asp.net/blob/dev/src/live.asp.net/Controllers/HomeController.cs#L24

File: src/live.asp.net/Services/FeedService.cs
Comment (hashhar) [reply src/live.asp.net/Services/FeedService.cs]: Oh, I understand now. Okay. I'll try implementing and pushing it.

File: src/live.asp.net/Services/FeedService.cs
Comment (campersau) [inline src/live.asp.net/Services/FeedService.cs]: According to the [Atom spec](https://tools.ietf.org/html/rfc4287) dates should be formatted with [`yyyy-MM-ddTHH:mm:ssK`](https://tools.ietf.org/html/rfc3339)

Comment (hashhar) [reply src/live.asp.net/Services/FeedService.cs]: Nice catch. Thanks, I overlooked it.

File: src/live.asp.net/Services/FeedService.cs
Comment (hishamco) [inline src/live.asp.net/Services/FeedService.cs]:
```csharp
w.WriteElementString("title", "ASP.NET Community Standup Episodes");
```

File: src/live.asp.net/Services/FeedService.cs
Comment (hishamco) [inline src/live.asp.net/Services/FeedService.cs]: Rename the variable

File: src/live.asp.net/Views/Shared/_Layout.cshtml
Comment (hishamco) [inline src/live.asp.net/Views/Shared/_Layout.cshtml]: Subscribe is more generic term!!

**Comment by hashhar** [comment]
Only the POCO part remains now. Other than that everything works. I also support so that podcast apps can use the videos too.

File: src/live.asp.net/Controllers/HomeController.cs
Comment (hishamco) [inline src/live.asp.net/Controllers/HomeController.cs]: I was planning to tell you about that 😄

File: src/live.asp.net/Views/Shared/_Layout.cshtml
Comment (hishamco) [inline src/live.asp.net/Views/Shared/_Layout.cshtml]: Same as above

File: src/live.asp.net/Views/Shared/_Layout.cshtml
Comment (hishamco) [inline src/live.asp.net/Views/Shared/_Layout.cshtml]: Episodes

File: src/live.asp.net/Services/YouTubeShowsService.cs
Comment (hishamco) [inline src/live.asp.net/Services/YouTubeShowsService.cs:218]: 👍

**Comment by hashhar** [comment]
@hishamco @campersau This is finished. Nothing more to do here. Should I squash the commits or leave them as is?

Also, the empty default constructors are there only for the `XMLSerializer` to work.

File: src/live.asp.net/Services/FeedService.cs
Comment (hishamco) [inline src/live.asp.net/Services/FeedService.cs]: It will be nice if we named `AtomFeed` because `Feed` is a generic term

File: src/live.asp.net/Services/FeedService.cs
Comment (hishamco) [inline src/live.asp.net/Services/FeedService.cs]: Use property instead, and move the attribute above

File: src/live.asp.net/Services/FeedService.cs
Comment (hishamco) [inline src/live.asp.net/Services/FeedService.cs]: Same as previous comment

File: src/live.asp.net/Services/FeedService.cs
Comment (hishamco) [inline src/live.asp.net/Services/FeedService.cs]: You can simplify this by adding a property in `Entry` class

**Review by hishamco — CHANGES_REQUESTED** [review]
Can you please refer to this https://validator.w3.org/feed/docs/atom.html to avoid unnecessary classes

**Comment by hashhar** [comment]
Pushed new changes.
- Removed optional elements.
- Uses properties instead of fields.
- Got rid of extra constructors.
- Conforms completely to the W3C Feed Validator now.
- Got rid of xsi and xsd specifications from root element.

File: src/live.asp.net/Controllers/HomeController.cs
Comment (hishamco) [inline src/live.asp.net/Controllers/HomeController.cs]: Add new line before this as part of code formatting standard

File: src/live.asp.net/Services/FeedService.cs
Comment (hishamco) [inline src/live.asp.net/Services/FeedService.cs]: Please add new line between properties

File: src/live.asp.net/Services/FeedService.cs
Comment (hishamco) [inline src/live.asp.net/Services/FeedService.cs]: Formatting should applied everywhere

File: src/live.asp.net/Services/FeedService.cs
Comment (hishamco) [inline src/live.asp.net/Services/FeedService.cs]: New line

**Review by hishamco — CHANGES_REQUESTED** [review]
Can you please refer to this https://blogs.msdn.microsoft.com/dancre/2004/12/13/generating-an-rss-feed-with-xmlserializer-asp-net-and-c/ because I think there 're many POCOs that can replaced by properties

**Comment by hashhar** [comment]
@hishamco I think I can't reduce POCOs because any XML Element that acts as a container for other elements needs to have it's own class since the `XmlRoot` annotation can only be applied to classes, strutcs, enums or returns. Something like `<a>` in the example below will need its own class as far as I know.

```
<a>
    <b foo="foo"/>
    <c bar="bar"/>
</a>
```

**Comment by hishamco** [comment]
@hashhar did you check the link I posted before?!!

**Comment by hashhar** [comment]
@hishamco Yes I read it. The problem is that I have three extra POCOs due to `<author>`, `<content>` and `<link>` tags. This is the minimum due to ATOM specification. But yes, I can get rid of a lot of these annotations by changing the variable names. I didn't know that.

**Comment by hishamco** [comment]
I think you're right, my aim to remove unnecessary classes, but after I looked [SyndicationItem](https://referencesource.microsoft.com/#System.ServiceModel/System/ServiceModel/Syndication/SyndicationItem.cs,899eb7ace8d2ea2a) there are Link and Person and Content classes, it would be nice if you check this and come up with a beauty piece of code 😄 
thanks

**Comment by hashhar** [comment]
@hishamco Wow, that's so good. I didn't know there was a reference source. I'll definitely take a look. But it may take some time since I have an exam tomorrow and the day after that.

I think this effort can be expanded and later ported over to ASP.NET Core. 😄

**Comment by hishamco** [comment]
The reference source is a great source of .NET :), regarding the Syndication Stuff, many of the previous APIs planned to come to the new Core world in upcoming NET STANDARDS

**Comment by hishamco** [comment]
FYI if we go with the required elements in the Atom feed we are able to remove unnecessary POCOs, so the remaining will be `AtomFeed` and `Entry`

File: src/live.asp.net/Services/FeedService.cs
Comment (hishamco) [inline src/live.asp.net/Services/FeedService.cs]: You can use
```csharp
Updated = show.ShowDate.DateTime.ToString("s") + "Z";
```

File: src/live.asp.net/Services/FeedService.cs
Comment (hishamco) [inline src/live.asp.net/Services/FeedService.cs]: I think `Entries` make sense as property name

**Comment by hashhar** [comment]
I simplified it the best I could. I removed the extra annotations where I could because by default XmlElement names are the same as variable names.
I also made the change to DateTime as you suggested and converted Link and Entry to links and entries to make sense because they are Collections.

File: src/live.asp.net/Services/FeedService.cs
Comment (hishamco) [inline src/live.asp.net/Services/FeedService.cs]: I think you break the naming conversions for the property again!!

Comment (hashhar) [reply src/live.asp.net/Services/FeedService.cs]: Hmmm, properties have to be uppercase only?

**Comment by hishamco** [comment]
Please have a look to my last blog [here](http://هشام.com/%D8%A5%D9%86%D8%B4%D8%A7%D8%A1-%D8%AA%D9%84%D9%82%D9%8A%D9%85-%D8%A7%D9%84%D9%88%D9%8A%D8%A8-atom-%D9%81%D9%8A-%D8%A7%D9%84%D9%80-asp.net-%D8%A8%D8%A5%D8%B3%D8%AA%D8%AE%D8%AF%D8%A7%D9%85-%D8%A7%D9%84%D9%80-xmlserializer) I know it's an Arabic 😄 , at least you can look to the code snippets, this is the less amount code I come up with to generate an ATOM feed, hope you like it

File: src/live.asp.net/Services/FeedService.cs
Comment (hashhar) [reply src/live.asp.net/Services/FeedService.cs]: Because making them lowercase allows me to remove the annotations where not needed. I can simply revert the commit if you want.

Comment (JimiC) [reply src/live.asp.net/Services/FeedService.cs]: It's just a standard property naming convention .NET/C# has for years.

Comment (hishamco) [reply src/live.asp.net/Services/FeedService.cs]: I know what you gonna say, but always refer to [CONTRIBUTING.md](https://github.com/aspnet/live.asp.net/blob/dev/CONTRIBUTING.md)

BTW what you're doing is good as contributor but you should follow the guidance, I suggest to finish everything and squash the commits if it's possible, after you can wait for **Damian Edwards** approval because he is the man behind the entire stuff in live.asp.net

Comment (hashhar) [reply src/live.asp.net/Services/FeedService.cs]: Okay. Thanks. I read the document and will change accordingly. Sorry for not paying attention to it early on.

**Comment by hashhar** [comment]
I squashed it all into a single commit. I don't think I can trim it down any more. Changing it to what you have in your blog will break two things.

1. In the current form, the content contains a thumbnail.
2. The media enclosure links allow feed readers to treat this as a podcast and will allow downloading the video.

Should I add doc comments to public functions?

**Comment by hashhar** [comment]
I just now tested with my own API key and it works without any issues. See the result.

Image

**Comment by hishamco** [comment]
I didn't mean that you should follow what I did, but what i'm doing is the least amount of code to generate the ATOM feed
I will try to test the code in my end, and let you know if there's any further suggestion, but in general that's great stuff 👍

**Pull request closed without merging**
